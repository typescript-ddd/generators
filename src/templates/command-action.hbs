import { CommandBus, CommandAction } from "@typescript-ddd/core";
import { {{commandName}}Command } from "../../infrastructure";

/**
 * Represents {{noun actionName}} command action. 
 */
export class {{pascal actionName}}CommandAction implements CommandAction {
  method: string = "{{method}}";
  path: string = "{{path}}";

  /**
   * Initializes a new instance of {{noun actionName}} command action.
   * @param {CommandBus} commandBus - The command bus.
   */
  constructor(private readonly commandBus: CommandBus) {}

  /**
   * Executes {{noun actionName}} command action.
   * @returns {Promise<void>}
   */
  public async execute(req: any, res: any): Promise<void> {
    {{#if requestParams}}
    {{#each requestParams}}
    const {{name}} = req.params.{{name}};
    {{/each}}
    const value = await this.commandBus.execute(new {{commandName}}Command(
      {{#each requestParams}}
      {{name}},
      {{/each}}
    ));
    {{else}}
    const value = await this.commandBus.execute(new {{commandName}}Command());
    {{/if}}
    {{#if status}}
    res.status({{status}}).json(value);
    {{else}}
    res.status(200).json(value);
    {{/if}}
  }
}
