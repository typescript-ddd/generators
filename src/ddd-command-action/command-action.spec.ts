import { CommandActionTemplateData } from "./command-action.types";
import { generateCommandAction } from "./command-action.generator";

describe("CommandAction", () => {

  it("should render template", () => {
    // Arrange
    const data: CommandActionTemplateData = {
      method: "POST",
      path: "/users",
      actionName: "PostUser",
      commandName: "CreatUser",
    };

    // Act
    const result = generateCommandAction(data);

    // Assert
    expect(result).toMatchSnapshot();
  });

  it("should render template with request params", () => {
    // Arrange
    const data: CommandActionTemplateData = {
      method: "POST",
      path: "/users",
      actionName: "PostUser",
      commandName: "CreatUser",
      requestParams: [
        { name: "name", type: "string" },
        { name: "email", type: "string" },
      ],
      status: 201,
    };

    // Act
    const result = generateCommandAction(data);

    // Assert
    expect(result).toMatchSnapshot();
  });
});
