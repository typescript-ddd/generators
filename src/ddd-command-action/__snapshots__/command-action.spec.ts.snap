// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandAction should render template 1`] = `
"import { CommandBus, CommandAction } from "@typescript-ddd/core";
import { CreatUserCommand } from "../../infrastructure";

/**
 * Represents a post user command action. 
 */
export class PostUserCommandAction implements CommandAction {
  method: string = "POST";
  path: string = "/users";

  /**
   * Initializes a new instance of a post user command action.
   * @param {CommandBus} commandBus - The command bus.
   */
  constructor(private readonly commandBus: CommandBus) {}

  /**
   * Executes a post user command action.
   * @returns {Promise<void>}
   */
  public async execute(req: any, res: any): Promise<void> {
    const value = await this.commandBus.execute(new CreatUserCommand());
    res.status(200).json(value);
  }
}
"
`;

exports[`CommandAction should render template with request params 1`] = `
"import { CommandBus, CommandAction } from "@typescript-ddd/core";
import { CreatUserCommand } from "../../infrastructure";

/**
 * Represents a post user command action. 
 */
export class PostUserCommandAction implements CommandAction {
  method: string = "POST";
  path: string = "/users";

  /**
   * Initializes a new instance of a post user command action.
   * @param {CommandBus} commandBus - The command bus.
   */
  constructor(private readonly commandBus: CommandBus) {}

  /**
   * Executes a post user command action.
   * @returns {Promise<void>}
   */
  public async execute(req: any, res: any): Promise<void> {
    const name = req.params.name;
    const email = req.params.email;
    const value = await this.commandBus.execute(new CreatUserCommand(
      name,
      email,
    ));
    res.status(201).json(value);
  }
}
"
`;
